feat(column): add :redrawstatuscolumn to rebuild 'statuscolumn'

luukvbaal opened the pull request:
Problem:  No obvious way to force a 'statuscolumn' rebuild.
Solution: Let :redraw force a 'statuscolumn' rebuild.

zeertzjq commented:
It can already be forced by setting 'statuscolumn' to itself.

zeertzjq commented:
Also if `:redraw` forces update for current window, then `:redraw!` should force update for all windows.

luukvbaal commented:
> It can already be forced by setting 'statuscolumn' to itself.

Are you saying it is possible to make the added test pass on master? I couldn't find a way to, I'll look at it again later. Even so, I think having `:redraw` force a rebuild makes sense.

zeertzjq commented:
Depends on what you mean by "pass on master". If you change `:redraw!` to `:let &stc = &stc` then it "passes on master".

luukvbaal commented:
Right, perhaps I typo'd when testing that. 

So do we update the docs to recommend setting `'stc'` to itself or go ahead with modifying `:redraw`?

luukvbaal committed:
feat(column): add :redrawstatuscolumn to rebuild 'statuscolumn'

Problem:  No obvious way to force a 'statuscolumn' rebuild.
Solution: Add :redrawstatuscolumn to force a 'statuscolumn' rebuild.

hbiel commented:
Thanks for digging into this!

Using `vim.o.statuscolumn = vim.o.statuscolumn` works perfectly in my case.

From a user perspective this _feels_ kinda hacky though. A dedicated command, like maybe `:redrawStatuscolumn`, would be more straightforward in my opinion. This could even just reset the statuscolumn to itself under the hood.

But keeping it as is would be fine by me too. Mentioning this in the docs would be great then.

luukvbaal commented:
@zeertzjq care to elaborate on what exactly requires discussion?
I have no immediate need for this but I think it's useful/expected that 'statuscolumn' will build from scratch on `:redraw`. And it's relatively free in the sense that if this changes the 'statuscolumn' width, the entire window would already be redrawn anyways.

zeertzjq commented:
`:redraw` "redraws **pending** screen updates", so "build from scratch" is definitely not "expected".

luukvbaal commented:
> `:redraw` "redraws **pending** screen updates", so "build from scratch" is definitely not "expected".

Yes, that's what I said in my edit. Thoughts on doing this in `redrawstatus`?

zeertzjq commented:
I think that makes more sense.

luukvbaal commented:
My last attempt at this, if a new command is deemed unwarranted, I guess we just add the workaround to the docs. 

Pairing a rebuild of the statuscolumn with `:redrawstatus` seems less than ideal, because it is much more expensive than just redrawing the status lines.

The `vim.o.statuscolumn = vim.o.statuscolumn` workaround is less than ideal because it does unnecessary option value/statusline-format checking.

clason commented:
> The vim.o.statuscolumn = vim.o.statuscolumn workaround is less than ideal because it does unnecessary option value/statusline-format checking.

Yes, and we don't want to further propagate such hacks (`&bg=&bg`, anyone)?

If a custom Vimscript command is not acceptable (and we've long had the policy of "vimscript interface belongs to vim; we don't make changes"), we might want to think about a Lua API `vim.redraw(opts)` that allows for more fine-grained control.

luukvbaal commented:
Yeah `vim.redraw()` might be nice to have, I would be willing to take a stab at that in place of this PR if a new command is unwanted. But I think in this case `:redrawstatuscolumn` helps with discoverability next to `:redraw/status/tabline`.

luukvbaal commented:
> > Pairing a rebuild of the statuscolumn with :redrawstatus seems less than ideal, because it is much more expensive than just redrawing the status lines.
> 
> Is this really a problem in practice?

I think `:redrawstatus` is used too often (by plugins) to have it force redraw the entire screen when `'statuscolumn'` is set. The situation this new command addresses seems quite niche, I think it shouldn't affect the common case.
